src/Fmt_ast.ml

```diff
@@ -186,6 +186,22 @@ let wrap_exp_if c ~parens ~loc k =
   | `Parens -> wrap_if parens "(" ")" k
   | `Begin_end -> wrap_fits_breaks_exp_begin_end ~parens k
 
+let wrap_disambiguate_exp_if c ~parens ~loc k =
+  match parens_or_begin_end c ~loc with
+  | `Parens ->
+      if c.conf.disambiguate_non_breaking_match then
+        wrap_if_fits_or parens "(" ")" k
+      else wrap_if parens "(" ")" k
+  | `Begin_end -> wrap_fits_breaks_exp_begin_end ~parens k
+
+let wrap_disambiguate_fits_breaks_exp_if ?(space = true) c ~parens ~loc k =
+  match parens_or_begin_end c ~loc with
+  | `Parens ->
+      if c.conf.disambiguate_non_breaking_match then
+        wrap_if_fits_or parens "(" ")" k
+      else wrap_fits_breaks_if ~space c.conf parens "(" ")" k
+  | `Begin_end -> wrap_fits_breaks_exp_begin_end ~parens k
+
```

Comment:
I am not very happy to add yet another layer to the wrap functions, it's very messy, and I'm not sure it would be better to have a single wrap function parameterized by many booleans (disambiguate, fits/breaks, preserve begin/end, the parens condition, etc.)

Reply:
Indeed looks better.